R/tree.R

```diff
@@ -0,0 +1,314 @@
+#' Pretty tree-like object printing
+#'
+#' A cleaner and easier to read replacement for `str` for nested list-like
+#' objects
+#'
+#' @param x A tree like object (list, etc.)
+#' @param index_arraylike Should children of containers without names have
+#'   indices used as stand-in?
+#' @param max_depth How far down the tree structure should be printed. E.g. `1`
+#'   means only direct children of the root element will be shown. Useful for
+#'   very deep lists.
+#' @param val_printer Function that values get passed to before being drawn to
+#'   screen. Can be used to color or generally style output.
+#' @param class_printer Same as `val_printer` but for the the class types of
+#'   non-atomic tree elements.
+#' @param show_attributes Should attributes be printed as a child of the list or
+#'   avoided?
+#' @param remove_newlines Should character strings with newlines in them have the
+#'   newlines removed? Not doing so will mess up the vertical flow of the tree
+#'   but may be desired for some use-cases if newline structure is important to
+#'   understanding object state.
+#' @param
+#' char_vertical,char_horizontal,char_branch,char_final_branch,char_vertical_attr,char_horizontal_attr
+#' Unicode characters used to construct the tree. Typically you wont want to
+#' change these.
+#' @param ... Ignored (used to force use of names)
+#'
+#' @return console output of structure
+#'
+#' @examples
+#'
+#' x <- list(
+#' list(id = "a",
+#'      val = 2),
+#' list(id = "b",
+#'      val = 1,
+#'      children = list(
+#'        list(id = "b1",
+#'             val = 2.5),
+#'        list(id = "b2",
+#'             val = 8,
+#'             children = list(
+#'               list(id = "b21",
+#'                    val = 4)
+#'             )))),
+#' list(id = "c",
+#'      val = 8,
+#'      children = list(
+#'        list(id = "c1"),
+#'        list(id = "c2",
+#'             val = 1))))
+#'
+#' # Basic usage
+#' tree(x)
+#'
+#' # Even cleaner output can be achieved by not printing indices
+#' tree(x, index_arraylike = FALSE)
+#'
+#' # Limit depth if object is potentially very large
+#' tree(x, max_depth = 2)
+#'
+#' # You can customize how the values and classes are printed if desired
+#' tree(x, val_printer = function(x){paste0("_", x, "_")})
+#'
+#' @export
+tree <- function(
+  x,
+  ...,
+  index_arraylike = TRUE,
+  max_depth = Inf,
+  val_printer = crayon::blue,
+  class_printer = crayon::silver,
+  show_attributes = FALSE,
+  remove_newlines = TRUE,
+  char_vertical = "\u2502",
+  char_horizontal = "\u2500",
+  char_branch = "\u251c",
+  char_final_branch = "\u2514",
+  char_vertical_attr = "\u250A",
+  char_horizontal_attr = "\u2504"
+){
+  args_in_dots <- list(...)
```

Comment:
Can you please just use `ellipsis::check_dots_empty()`? (This will be inlined into rlang in the future so no need to worry about the extra dep)